servlet.py

```diff
@@ -438,6 +438,109 @@ def get(self):
                                            deformat=deformat, reformat=reformat)
 
 
+class TranslateChainHandler(TranslateHandler):
+
+    pairs_graph = {}
+    path_cache = {}
+
+    def initPairsGraph(self):
+        for pair in self.pairs:
+            l1, l2 = pair.split('-')
+            if l1 in self.pairs_graph:
+                self.pairs_graph[l1].append(l2)
+            else:
+                self.pairs_graph[l1] = [l2]
+
+    def shortestPath(self, start, end):
+        if (start, end) in self.path_cache:
+            return self.path_cache[(start, end)]
+        queue = [[start]]
+        visited = set()
+        while queue:
+            path = queue.pop(0)
+            node = path[-1]
+            if node == end:
+                self.path_cache[(start, end)] = path
+                return path
+            if node not in visited:
+                if node != start:
+                    self.path_cache[(start, node)] = path
+                for adjacent in self.pairs_graph.get(node, []):
+                    new_path = list(path)
+                    new_path.append(adjacent)
+                    queue.append(new_path)
+                visited.add(node)
+
+    def pairList(self, lgs):
+        return [(lgs[i], lgs[i+1]) for i in range(0, len(lgs)-1)]
+
+    def getPairsOrError(self, langpairs, text_length):
```

Comment:
The code is very similar, but everywhere `getPairOrError()` uses a single pair this uses a list, so doing it with inheritance would be hard. If the two handlers were merged this would still need to be a separate method.